Source/Session.swift

```diff
@@ -513,41 +513,35 @@ extension Session: RequestDelegate {
         return session.configuration
     }
 
-    public func willAttemptToRetryRequest(_ request: Request) -> Bool {
-        return retrier(for: request) != nil
-    }
-
-    public func retryRequest(_ request: Request, ifNecessaryWithError error: Error) {
-        guard let retrier = retrier(for: request) else { request.finish(); return }
+    public func retryResult(for request: Request, dueTo error: Error, completion: @escaping (RetryResult) -> Void) {
+        guard let retrier = retrier(for: request) else {
+            rootQueue.async { completion(.doNotRetry) }
```

Comment:
Any particular reason is async? It seems like you could just immediately call the completion handler, since `retryResult` should always be called on the root queue, right?

Reply:
Yep, you're absolutely right, you could. I chose to make it async to make the behavior consistent between both cases. If you call the `completion` closure on the calling thread, then the completion will run at the call-site before the next line of local scope is executed. In actual async cases, the local scope logic is executed before the completion closure is called. As a general guiding principle, it is wise to always call completion closures in either a consistent synchronous or asynchronous case to ensure client logic executes in a consistent manner.

Reply:
👍